---
title: "Finding markers with Seurat"
author: "Sonny"
output: html_document
---

This notebook/script places after "Loading_AIBS_Data.Rmd". We are now finding cell subclass markers with Seurat and the AIBS data. 
As implied, we assume that Seurat (V3+) is installed.
First, we convert the data into a Seurat object

```{r, data load and prep in Seurat}

library(Seurat) #load the needed library

tome_sample_meta_filtered$query_subclass <- paste(tome_sample_meta_filtered$split_class, tome_sample_meta_filtered$split_subclass, sep="_") #generate a final set of labels for subclass by combining class labels and subclass gene labels

unique(tome_sample_meta_filtered$query_subclass) #these will be our subclasses/key cell-group identifies that we use for getting cell markers

Seu_AIBS_obj <- CreateSeuratObject(counts = AIBS_Rawcount_InEx, meta.data = tome_sample_meta_filtered) #after confirming our grouping variable, this creates the Seurat object

# check that the seurat object data matches what we provided as input:

test <- as.data.frame(Seu_AIBS_obj@assays$RNA@counts) #extracting count matrix from Seurat object
identical(test, AIBS_Rawcount_InEx) #indentical check

test <- as.data.frame(Seu_AIBS_obj@meta.data) #extracting metadata from Seurat object
test <- test[,4:length(colnames(test))] #removing extra columns made by Seurat
identical(test, tome_sample_meta_filtered) #identical test

remove(test)

# finally, we get log cpm values of the data

Seu_AIBS_obj <- NormalizeData(Seu_AIBS_obj, normalization.method = "LogNormalize", scale.factor = 1000000)

Seu_AIBS_obj[["RNA"]]@data #see our normalized data

```

Sometimes, at this stage (or to do a new analysis on slightly-altered Seurat object), you may realize you want to further change your data in the Seurat. You can do this without going back to the metadata df or the count matrix with Seurat functions.

```{r, modifying Seurat object before finding markers}

Idents(Seu_AIBS_obj) <- "query_subclass" #apply/assign our identity labels of interest to cells (we would just do this if there were no modifications to be made to the Seurat object)

### somtimes, we might want to change the way we group our cells

Seu_AIBS_obj$it2_query_subclass <- Idents(Seu_AIBS_obj) #making a new identity variable; starting by copying the previous active idents (which was "query_subclass" at the time); this needs the first command in this chunk (line 45) to be run first; we are effectively making a copy of "query_subclass" here

levels(Seu_AIBS_obj$it2_query_subclass)[levels(Seu_AIBS_obj$it2_query_subclass)=="Oligo_MOBP"] <- "Oligo" #collapsing the oligo groups
levels(Seu_AIBS_obj$it2_query_subclass)[levels(Seu_AIBS_obj$it2_query_subclass)=="Oligo_OPALIN"] <- "Oligo" #collapsing the oligo groups

levels(Seu_AIBS_obj$it2_query_subclass)[levels(Seu_AIBS_obj$it2_query_subclass)=="Exc_RORB"] <- "Pyramidal" #collapsing the exc groups
levels(Seu_AIBS_obj$it2_query_subclass)[levels(Seu_AIBS_obj$it2_query_subclass)=="Exc_LINC00507"] <- "Pyramidal" #collapsing the exc groups
levels(Seu_AIBS_obj$it2_query_subclass)[levels(Seu_AIBS_obj$it2_query_subclass)=="Exc_FEZF2"] <- "Pyramidal" #collapsing the exc groups
levels(Seu_AIBS_obj$it2_query_subclass)[levels(Seu_AIBS_obj$it2_query_subclass)=="Exc_THEMIS"] <- "Pyramidal" #collapsing the exc groups

Idents(Seu_AIBS_obj) <- "it2_query_subclass" #setting the active identity (the grouping variable that Seurat will use to distinguish between groups of cells, for calculating cell markers) to the new set of identity variables

### sometimes, we might want to further filter out some cells from our analyses

Seu_AIBS_obj <- subset(Seu_AIBS_obj, idents = c("Inh_ADARB2", "Inh_LHX6"), invert = TRUE) #we want all cells except for these two groups

```

The collapsing of the group names is to help us compare to a reference marker list later on (see Marker_evaluation notebook for more details).
The removal of interneuron clusters ADARB2 and LHX6 is due to our evaluation of the information from https://celltypes.brain-map.org/rnaseq/human_ctx_smart-seq , which suggests that these two groups/markers may not clearly delineate from the others in our analysis.

We are now ready to find cell markers.
To find markers (for cell subclass), we assign the appropriate identity to the cells/samples, and calculate markers by using the FindAllMarkers function. There are many arguments/thresholds we can use in this case, and to determine the best thresholds to use, we may want to check on the characteristic metrics (such what percentage of a group of cells express a gene) of key, cluster-defining markers. For our (demonstration) purposes, our key markers are LAMP5, VIP, SST, SLC17A7, and PVALB.

```{r, find markers}

Idents(Seu_AIBS_obj) <- "query_subclass" #apply/assign our identity labels of interest to cells
Idents(Seu_AIBS_obj) <- "it2_query_subclass" #apply/assign our modified/edited identity labels instead, if desired (we do, for this example)

### there are multiple options to find markers initially

# do all comparisons, using defaults (Wilcoxon rank-sum test)
Seu_AIBS_markers <- FindAllMarkers(object = Seu_AIBS_obj) # if computational resources/time is not an issue, using all default settings

# if time/resources are an issue, you may need to find markers for each group separately, such as by the example code below:
Seu_AIBS_SST_markers <- FindMarkers(Seu_AIBS_obj, ident.1 = "Oligo_OPALIN") #this is not run for the generation of this notbook

# however, for us, we want to use MAST for single-cell datasets

BiocManager::install("MAST")
library("MAST")

Seu_AIBS_markers_screen <- FindAllMarkers(Seu_AIBS_obj, logfc.threshold = .1, min.pct = .1, only.pos = TRUE, return.thresh = .05, test.use = "MAST") # we use this as a dummy run; we don't care about the results per se, but rather the fold-changes and detected percentages of key markers in their cannonical groups (e.g: SST in SST cluster) 

write.csv(Seu_AIBS_markers_screen, "/external/rprshnas01/kcni/ychen/git/MarkerSelection/Data/Outputs/lenient_ACC_results.csv") #save results

# finding the most lenient parameters that would allow us to detect the key markers (derived from the results of Seu_AIBS_markers_liberal), we can rerun the test with these parameters

Seu_AIBS_markers <- FindAllMarkers(Seu_AIBS_obj, logfc.threshold = 2.5, min.pct = .35, only.pos = TRUE, return.thresh = .05, test.use = "MAST")

```

The results (when using FindAllMarkers) are in one dataframe in a "gathered" format (a column for your queried cell group) with all marker genes. The default results also have adjusted p values ranging from 1 to 0 (or near to it).

```{r, preliminary look through results}

head(Seu_AIBS_markers) #a quick look at the format of the dataframe

#we can filter these results, such as by adjusted p-value

Seu_AIBS_markers_filtered <- Seu_AIBS_markers[Seu_AIBS_markers$p_val_adj < .05,] #p value filter

head(Seu_AIBS_markers_filtered) #sample view of top results in dataframe

```

However, the above is just a general, one-pass attempt at getting markers using default settings in Seruat. Often times, the process will need to be repeated and reinterated to fine-tune the options and cell-grouping resolution used. The below is an example of modifying some parameters to try again.

```{r, a reiteration with changed parameters}

# trying different threshold parameters based on AIBS workflow, brought to our attention by Jordan Sicherman 

Seu_AIBS_markers_stringent <- FindAllMarkers(object = Seu_AIBS_obj_ACC_it2, slot = "data", logfc.threshold = 1, test.use = "MAST", min.pct = .7, min.diff.pct = .8, only.pos = TRUE, return.thresh = .05) 

# trying a different test for comparing gene expression between different groups, also usable in Seurat

Seu_AIBS_markers_roc <- FindAllMarkers(Seu_AIBS_obj_ACC_it2, logfc.threshold = 2.5, min.pct = .35, only.pos = TRUE, return.thresh = .05, test.use = "roc")

```

We can also try finding markers by systematically doing pairwise comparisons between all groups.

```{r, pairwise}

paircomp_groups <- unique(Idents(Seu_AIBS_obj_ACC_it2)) #get list of all groups
query_group <- "Inh_SST" #set group of interest
paircomp_groups = paircomp_groups[!paircomp_groups %in% query_group] #remove group of interest from list of cell groups

Seu_AIBS_pair_ACC_SST <- data.frame(p_val=double(),
                 avg_logFC=double(),
                 pct.1=double(),
                 pct.2=double(),
                 p_val_adj=double(),
                 group_2=character(),
                 stringsAsFactors=FALSE) #initialize blank result df

for (cluster_name in paircomp_groups) {

  temp_df <- FindMarkers(object = Seu_AIBS_obj_ACC_it2, ident.1 = query_group, ident.2 = cluster_name, logfc.threshold = 2.5, min.pct = .35, only.pos = TRUE, test.use = "MAST") #get results for pairwise comparison between group of interest and current group in loop
  temp_df$group_2 <- cluster_name #add new column indicating the comparator/group 2
  Seu_AIBS_pair_ACC_SST <- rbind(Seu_AIBS_pair_ACC, temp_df) #add results to overall result df

}
 
write.csv(Seu_AIBS_pair_ACC_SST, "/external/rprshnas01/kcni/ychen/git/MarkerSelection/Data/Outputs/pairwise_ACC_SST_results.csv") #save results

```


This is the end of this notebook for now. We can now take this list of markers and evaluate them or use them for other analyses. For quicker iteration of this workflow, you can look at our iteration of this process looking at MTG neurons instead of ACC neurons in Code/MTG_analysis.R, though it is less documented/explained, it follows the same steps detailed in this notebook.
